Include cleanup in VideoPlayer, workaround for utils/RegExp.h

MaxKellermann opened the pull request:
See https://github.com/xbmc/xbmc/pull/9442 for discussion.

garbear commented:
jenkins build this please

FernetMenta commented on xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxUtils.cpp:
```diff
@@ -23,6 +23,7 @@
 #endif
 #include "DVDDemuxUtils.h"
 #include "DVDClock.h"
+#include "linux/XMemUtils.h"
```
Windows of course complains about this :)

MaxKellermann replied:
Oh, there was the only definition of _aligned functions I found, but I didn't know these were provided by the MSVC runtime.

MaxKellermann commented:
Ok, changed `XMemUtils.h` to `system.h`. Looks like `system.h` is the frontend for all system-specific headers?

fritsch commented:
system.h is a melting pot.

2016-03-30 9:57 GMT+02:00 Max Kellermann <notifications@github.com>:

> Ok, changed XMemUtils.h to system.h. Looks like system.h is the frontend
> for all system-specific headers?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/9494#issuecomment-203306859>
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

MaxKellermann commented:
Is that good or bad?
I think it's bad because it's a fat get-it-all thing, and I like very lean header dependencies.

fritsch commented:
Keep it - there are worse things to do. It's the way it was meant to be done without cluttering a whole lot of #ifdeffery in there ... one can think at a point to split system.h to system_core.h system_whatever.h to get stuff more fine granularly.

MaxKellermann commented:
Yes, splitting into many system-specific headers was what I had in mind. That's what I do in my other projects. If I know that all I need is `_aligned_malloc()`, I'd like to include just `system/AlignedMalloc.h` instead of the big beast.

Headers like `system.h` confuse me, because I never know what may fall apart when I remove it; it defines so many macros - what if some source depends on a macro defined by `system.h`, but doesn't actually include it, it just happens to be included by another header included by another header ...? I thinkt the concept of Kodi's `system.h` is horribly fragile.

fritsch commented:
@MaxKellermann yes - it has grown over the years and one added this, the others added that and so on ... another 5 cleanup cycles and all is fine.

notspiff commented:
system.h isn't that fat actually. main fat comes from PlatformInclude.h. in my fork PlatformInclude.h is gone. i wanted to submit those buts but it relies on https://github.com/xbmc/xbmc/pull/8798 and that has stalled.

MrMC commented:
system.h is historical and very, very sensitive to placement. Rule #1, it goes at the top, not bottom or in some intermediate place in the include order. Pretty much all the HAVE -> HAS and other critical defines for cross-platform are in system.h. This stuff has to go somewhere.

PlatformInclude.h and friends is another sensitive one. 

Moral, mess with either of this without thinking about cross-platform, and epic fail :)

As for _aligned_malloc, there are very few reasons to even do this. In Darwin, it's not needed as memory alllocs are already aligned.

fritsch commented:
For FFmpeg we use the ffmpeg internal malloc methods that have an alignment parameter, cause here SSE intrinsics are highly relevant. In VAAPI we even use 64 byte alignment as we are using simd stream extensions.

MaxKellermann commented:
On 2016/03/30 15:08, Scott Davilla <notifications@github.com> wrote:
> system.h is historical and very, very sensitive to placement. Rule
> #1, it goes at the top, not bottom or in some intermediate place in
> the include order.

This is exactly the kind of thing I don't like about `system.h`.  It's
fragile.  And the requirements you just named are not documented -
which is pretty bad for something that is so important.

There's no real reason why `system.h` would need to be at the top.
Anybody who depends on HAVE/HAS macros should just include it before
checking those macros.  Or better: have subsystem-specific "feature"
headers.  Like this one I wrote a few years ago:

 https://github.com/MaxKellermann/XCSoar/blob/master/src/Screen/OpenGL/Features.hpp

> As for _aligned_malloc, there are very few reasons to even do
> this. In Darwin, it's not needed as memory alllocs are already
> aligned.

So on Darwin, _aligned_malloc() should be an inline function mapping
right to malloc(), right?

But what happens if somebody happens to ask for an alignment that is
not provided by Darwin's defaults?  Let's say align to memory page?
"Already aligned" is up to the caller's interpretation.

Kodi doesn't need the full _aligned_malloc() API - all it needs a
function which allocates with an alignment suitable for the CPU's
vectoring instructions, doesn't it?  Using the low-level
_aligned_malloc() makes the code more complicated than it should be.

fritsch commented:
@MaxKellermann yes, you are right. Though cause of this fragility we can run into several issues - and even worse, we might only realize it later :-)

MrMC commented:
Darwin is default 16 byte aligned for malloc and friends. "The allocated memory is aligned such that it can be used for any data type, including AltiVec- and SSE-related types." 

There's also posix_memalign available.

int  posix_memalign(void **, size_t, size_t) __OSX_AVAILABLE_STARTING(__MAC_10_6, __IPHONE_3_0);

I suggest moving to posix_memalign which is found in stdlib.h for posix based platforms. Windows can emulate that.

Was added in Android at API level 16 but that needs checking.

in any case, _aligned_malloc should die :)

Just to say, a missing HAVE or HAS is the most difficult to track, if there are no build errors, you have to debug at runtime and typically it gets missed by developers and found by users when some feature no longer works or is missing. There are lots of features :)

Align to memory page, that means you are doing something very, very specific, most likely very platform dependent. That should never be in a general include header :)

MaxKellermann commented:
On 2016/03/30 16:45, Scott Davilla <notifications@github.com> wrote:
> Align to memory page, that means you are doing something very, very specific, most likely very platform dependent. That should never be in a general include header :)

Align for a CPU's vectoring instructions is platform dependent in any
case.  And it is very specific, used only by a few core algorithms
written in assembly or intrinsics.

In spite of that, _aligned_malloc() is provided by the global system.h
header to each and every Kodi source!

(The "page alignment" was just a random example to demonstrate that
"already aligned" is not well-defined, and your Darwin assumption is
dangerous; it depends on what the caller thinks it is, and every
caller may define it differently.)

MrMC commented:
Ah, my darwin assumption is not dangerous. I just know that platform very well. Apple goes to great lengths to guarantee malloc is safe for for any data type. That's why they state that very thing in their docs. That means asm, C, C++, ObjC, sysctrl calls, everything. They have been doing this for decades across many different cpu architectures. In over 20 years, I've never seen an malloc alignment error under darwin.

MaxKellermann commented:
On 2016/03/30 17:14, Scott Davilla <notifications@github.com> wrote:
> Ah, my darwin assumption is not dangerous.

I disagree.  It is dangerous!  You may know all about Apple and
vectoring, but that doesn't matter, because you forgot to consider the
"big unknown": a caller may pass a different alignment parameter to
_aligned_malloc(), for whatever reason we may not be able to imagine
right now.  If you just ignore the alignment parameter on Darwin
because you believe it doesn't matter on Darwin, _aligned_malloc()
doesn't work properly, and Kodi may segfault or kill your dog.

Fulfilling the API contract properly IS important.  Very important.

If you can't (or don't want to) fulfill the API contract properly, you
may change the API contract.  That was what I suggested: not allow
calling _aligned_malloc(), but wrap it in a simpler API that is
documented to align for the CPU's vectoring, and nothing more.

By limiting the scope of the API contract, you may apply your Darwin
optimization.

MrMC commented:
All I'm saying is the reason for doing an _aligned_malloc is very, very specific and not general. In darwin, there is not one public API that will fail using the results of their malloc, this includes asm and hw vector/floating point units. That is their API contract to developers. It is an overriding fact for all public darwin APIs. Once you start dipping into their private APIs, all bets are off, including that the API will even be consistent across platforms or version. That's why they are private, they can change at Apple whim :)

Now, I could probably think up some wacky thing that I might want page aligned or some other alignment but it would be very platform specific and in that case, I'd use posix_memalign to get me exactly what I needed. And that need would be heavily commented in code so that it is crystal clear why it is required.

FernetMenta commented:
jenkins build this please

FernetMenta commented:
@MaxKellermann jenkis was happy but for some reason it does not apply anymore. could you rebase please, then we can merge it.

MaxKellermann committed:
VideoPlayer: include cleanup

Reduce header dependencies.

MaxKellermann committed:
utils/Regexp: include stdlib.h before pcre.h

Including pcre.h inside the namespace breaks the system headers,
because pcre.h includes stdlib.h, and thus all stdlib.h definitions
now live in that namespace.

This did not break because Kodi's include order just happened to
include stdlib.h before util/Regexp.h.

By including stdlib.h manually right before opening the "PCRE"
namespace, we work around this systematic problem.  However, this
workaround is fragile, and moving pcre.h into a namespace is still an
ugly kludge.  At least this makes some obvious problems go away, and
allows us to clean up include dependencies/ordering.

MaxKellermann commented:
Rebased. Commit https://github.com/xbmc/xbmc/commit/3f6bbcb148c8a5f12fc36efc6847e5aa2310000f caused the conflict.

MartijnKaijser commented:
jenkins build this please

FernetMenta commented:
jenkins build and merge

MartijnKaijser commented:
Merge build don't work if one of the binary Addons is broken
